query-engine/core/src/query_graph_builder/write/create.rs

```diff
@@ -17,41 +18,47 @@ pub(crate) fn create_record(
     model: Model,
     mut field: ParsedField<'_>,
 ) -> QueryGraphBuilderResult<()> {
-    graph.flag_transactional();
-
     let data_map = match field.arguments.lookup(args::DATA) {
         Some(data) => data.value.try_into()?,
         None => ParsedInputMap::default(),
     };
 
-    let create_node = create::create_record_node(graph, query_schema, model.clone(), data_map)?;
-
-    // Follow-up read query on the write
-    let read_query = read::find_unique(field, model.clone())?;
-    let read_node = graph.create_node(Query::Read(read_query));
-
-    graph.add_result_node(&read_node);
-    graph.create_edge(
-        &create_node,
-        &read_node,
-        QueryGraphDependency::ProjectedDataDependency(
-            model.primary_identifier(),
-            Box::new(move |mut read_node, mut parent_ids| {
-                let parent_id = match parent_ids.pop() {
-                    Some(pid) => Ok(pid),
-                    None => Err(QueryGraphBuilderError::AssertionError(
-                        "Expected a valid parent ID to be present for create follow-up read query.".to_string(),
-                    )),
-                }?;
-
-                if let Node::Query(Query::Read(ReadQuery::RecordQuery(ref mut rq))) = read_node {
-                    rq.add_filter(parent_id.filter());
-                };
-
-                Ok(read_node)
-            }),
-        ),
-    )?;
+    if can_use_atomic_create(query_schema, &model, &data_map, &field) {
+        let create_node = create::atomic_create_record_node(graph, query_schema, model, data_map, field)?;
+
+        graph.add_result_node(&create_node);
+    } else {
+        graph.flag_transactional();
+
+        let create_node = create::create_record_node(graph, query_schema, model.clone(), data_map)?;
+
+        // Follow-up read query on the write
+        let read_query = read::find_unique(field, model.clone())?;
+        let read_node = graph.create_node(Query::Read(read_query));
+
+        graph.add_result_node(&read_node);
+        graph.create_edge(
+            &create_node,
+            &read_node,
+            QueryGraphDependency::ProjectedDataDependency(
+                model.primary_identifier(),
+                Box::new(move |mut read_node, mut parent_ids| {
+                    let parent_id = match parent_ids.pop() {
+                        Some(pid) => Ok(pid),
+                        None => Err(QueryGraphBuilderError::AssertionError(
+                            "Expected a valid parent ID to be present for create follow-up read query.".to_string(),
+                        )),
+                    }?;
+
+                    if let Node::Query(Query::Read(ReadQuery::RecordQuery(ref mut rq))) = read_node {
+                        rq.add_filter(parent_id.filter());
+                    };
+
+                    Ok(read_node)
+                }),
+            ),
+        )?;
```

Comment:
Reviewer's note: this is just a chunk of code that already existed, but wrapped in the `else` branch of a conditional